docs/sources.rst

```diff
@@ -100,6 +127,17 @@ will limit where the ansible playbook can run. A simple example looks like
 ``{"i": 2, "meta": {hosts: "localhost"}}``. ``hosts`` can be a comma delimited
 string or a list of host names.
 
+As the plugin have full access to an unbounded queue that is consumed by ansible-rulebbok
+we carefully recommend to use always the method ``asyncio.Queue.put`` to put events as it's a non-blocking call.
+To give free cpu cycles to the event loop to process the events, we recommend to use ``asyncio.sleep(0)``
+inmediately after the ``put`` method.
+
+.. note::
+    ansible-rulebook is intended to be a long running process and react to events over the time.
+    If the ``main`` function exits then the plugin and ansible-rulebook will be terminated.
+    Usually you may want to implement a loop that keeps running and waits for events endlessly.
```

Comment:
@Alex-Izquierdo I think we should add a line here that a Shutdown event is automatically inserted by ansible-rulebook when a plugin ends, which will cause the ansible-rulebook to end. If the ruleset uses multiple sources, each of the source has the ability to end, causing all sources to end. A running source will get canceled when the ansible-rulebook is ending. The rulebook can have actions which trigger a Shutdown, when that shutdown happens the running sources would be canceled. If you need logic to do some cleanup in your sources you should catch the exception  asyncio.CancelledError and do a cleanup. e.g. https://github.com/ansible/event-driven-ansible/blob/1f9ce9641361c96438146f5cb6bae10b60fb4464/plugins/event_source/webhook.py#L48

Reply:
Agree. 
- Empathize that if **only one**  source ends, everything ends. 
- Empathize that the shutdown action in the rulebook ends everything. 
- Mention to catch asyncio.CancelledError if the plugin needs to do cleanup.

Reply:
@mkanoor updated